Malikabh/pem parsing

malikabhi05 opened the pull request:
Temp PR

malikabhi05 committed:
Adding PEM support to tpm2_load

Signed-off-by: malikabhi05 <abhishek.malik@intel.com>

mayuanchenma committed:
tools/tpm2_sign and tpm2_verifysignature: fix sm2 sign and verifysignature bugs.

1. sm2 sign could not get output signature.
2. sm2 verify tss format signature failed.

Signed-off-by: mayuanchen <94815698+mayuanchenma@users.noreply.github.com>

Commit:
import: fix bug on using scheme

When scheme is specified in the template, the openssl load functions
clobber the scheme value and set it to TPM2_ALG_NULL. Only set the
algorithm to NULL if zero value is specified.

Fixes: #2997

Signed-off-by: William Roberts <william.c.roberts@intel.com>

Commit:
import: support restricted parents

Support a restricted parent with an aes128cfb symmetric parameter.

Fixes: #2998

Signed-off-by: William Roberts <william.c.roberts@intel.com>

Commit:
tpm2_alg_util: make cfb implies for asym detail

When creating an object with an algorithm specifier like rsa2048:aes
make that implied that it's aes128cfb for the objects asym detail.

Signed-off-by: William Roberts <william.c.roberts@intel.com>

idesai commented on tools/tpm2_load.c:
```diff
@@ -1,14 +1,25 @@
 
 #define MAX_SESSIONS 3
+
+#define PEM_BEGIN_CHARS "-----BEGIN"
```
Redundant. load the tsspem directly to determine pem formatting.

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+    return false;
+}
+
+char getChar(int value) {
+    if (value >=0 && value <= 9) {
+        return ((char) value) + '0';
+    } else if (value >= 10 && value <= 15) {
+        return ((char) value) + 'a';
+    } else {
+        LOG_ERR("Unable to create parent using create primary");
+        return 'Z';
+    }
+}
```
Use [this](https://github.com/tpm2-software/tpm2-tools/blob/3bf7a12a879251d52b302507da66fb7af33e20ab/lib/tpm2_util.h#L141) instead.

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+     }
+};*/
+
+static const TPM2B_PUBLIC primaryEccTemplate = {
+    .publicArea = {
+        .type = TPM2_ALG_ECC,
+        .nameAlg = TPM2_ALG_SHA256,
+        .objectAttributes = (TPMA_OBJECT_USERWITHAUTH |
+                             TPMA_OBJECT_RESTRICTED |
+                             TPMA_OBJECT_DECRYPT |
+                             TPMA_OBJECT_NODA |
+                             TPMA_OBJECT_FIXEDTPM |
+                             TPMA_OBJECT_FIXEDPARENT |
+                             TPMA_OBJECT_SENSITIVEDATAORIGIN),
+        .authPolicy = {
+             .size = 0,
+         },
+        .parameters.eccDetail = {
+             .symmetric = {
+                 .algorithm = TPM2_ALG_AES,
+                 .keyBits.aes = 128,
+                 .mode.aes = TPM2_ALG_CFB,
+              },
+             .scheme = {
+                .scheme = TPM2_ALG_NULL,
+                .details = {}
+             },
+             .curveID = TPM2_ECC_NIST_P256,
+             .kdf = {
+                .scheme = TPM2_ALG_NULL,
+                .details = {}
+             },
+         },
+        .unique.ecc = {
+             .x.size = 0,
+             .y.size = 0
+         }
+     }
+};
+
```
use getcap to determine available algorithms on the TPM. If ECC is supported then always use ECC as parent object, else use RSA as parent object. @williamcroberts do you agree?

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+     }
+};
+
+void print_parent_name(ESYS_CONTEXT *esys_ctx, ESYS_TR *parent) {
+    TPM2B_NAME *name = 0;
+    tool_rc rc = tpm2_tr_get_name(esys_ctx, *parent, &name);
+    if (rc != tool_rc_success) {
+        printf("Unable to get parent name\n");
+    }
+    printf("Name Before: ");
+    for (int i = 0; i < name->size; i++) {
+        printf("%x\n", name->name[i]);
+    }
+    printf("\n");
+}
+
```
debug left over - remove.

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+}
+
+tool_rc setup_primary(ESYS_CONTEXT *esys_ctx, ESYS_TR *parent) {
+    TSS2_RC r;
```
```suggestion

```

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+        return tool_rc_general_error;
+    }
+    
+    r = Esys_CreatePrimary(esys_ctx, ESYS_TR_RH_OWNER,
```
```suggestion
    TSS2_RC r = Esys_CreatePrimary(esys_ctx, ESYS_TR_RH_OWNER,
```

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+tool_rc setup_primary(ESYS_CONTEXT *esys_ctx, ESYS_TR *parent) {
+    TSS2_RC r;
+    const TPM2B_PUBLIC *primaryTemplate = NULL;
+    //ESYS_TR primary;
```
```suggestion

```

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+                            parent, NULL, NULL, NULL, NULL);
+    
+    //print_parent_name(esys_ctx, parent);
+    if (r == 0x000009a2) {
```
```suggestion
    if (r != TSS2_RC_SUCCESS) {
```

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+    return tool_rc_success;
+}
+
+void get_parent(TSSPRIVKEY_LOAD *tpk, char* ptr_const) {
+    int temp;
+    int j = 0;
+
+    ptr_const[j] = '0';
+    ptr_const[++j] = 'x';
+    for (int i = 0; i < tpk->parent->length; i++) {
+        temp = tpk->parent->data[i] >> 4;
+        ptr_const[++j] = getChar(temp);
+        temp = tpk->parent->data[i] & 0xf;
+        ptr_const[++j] = getChar(temp);
+    }
+    ptr_const[++j] = '\0';
+}
+
```
redundant -> use https://github.com/tpm2-software/tpm2-tools/blob/3bf7a12a879251d52b302507da66fb7af33e20ab/lib/tpm2_util.h#L141

idesai commented on tools/tpm2_load.c:
```diff
@@ -103,7 +316,8 @@ static tool_rc process_output(ESYS_CONTEXT *ectx) {
 
-    UNUSED(ectx);
+    //UNUSED(ectx);
+    char ptr_const[100];
```
```suggestion

```

idesai commented on tools/tpm2_load.c:
```diff
@@ -129,16 +337,87 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
+        input_bio = BIO_new_file(ctx.object.privpath, "rb");
+
+        // -- CODE BLOCK 1 --
+        TSSPRIVKEY_LOAD *tpk = NULL;
+        tpk = PEM_read_bio_TSSPRIVKEY_LOAD(input_bio, NULL, NULL, NULL);
```
```suggestion
        TSSPRIVKEY_LOAD *tpk = PEM_read_bio_TSSPRIVKEY_LOAD(input_bio, NULL, NULL, NULL);
        bool is_tss_pem = tpk ? true : false;
```

idesai commented on tools/tpm2_load.c:
```diff
@@ -129,16 +337,87 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
+            LOG_ERR("Unable to convert parent string to integer");
+            return tool_rc_general_error;
+        }
+        if (parent_val == TPM2_RH_OWNER || parent_val == 0) {
```
bool is_persistent_parent = (parent_val != TPM2_RH_OWNER && parent_val != 0)

idesai commented on tools/tpm2_load.c:
```diff
@@ -129,16 +337,87 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
+            LOG_ERR("Unable to get session handle for parent");
+            return tool_rc_general_error;
+        }
+        ctx.parent.object.session = sess_handle;
```
check if this value can be assigned directly as ESYS_TR_PASSWORD

idesai commented on tools/tpm2_load.c:
```diff
@@ -172,20 +451,24 @@ static tool_rc check_options(ESYS_CONTEXT *ectx) {
         LOG_ERR("Expected private object portion via -r");
         rc = tool_rc_option_error;
     }
+    // check for PEM file
+    // if not PEM then we need to make sure we check everything
+    if (!is_PEM(ctx.object.privpath)) {
+        if (!ctx.parent.ctx_path) {
+            LOG_ERR("Expected parent object via -C or a PEM file with -r");
+            rc = tool_rc_option_error;
+        }
+
+        if (!ctx.object.pubpath) {
+            LOG_ERR("Expected public object portion via -u or a PEM file with -r");
+            rc = tool_rc_option_error;
+        }
+    }
```
bool is_tsspem =(!ctx.parent.ctx_path && !ctx.object.pubpath)

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+
+    primaryTemplate = &primaryEccTemplate;
+
+    if(!primaryTemplate) {
```
```suggestion
    if(!primary_template) {
```
no camel casing please

idesai commented on tools/tpm2_load.c:
```diff
@@ -54,6 +89,184 @@ static tool_rc load(ESYS_CONTEXT *ectx) {
+    }
+    
+    r = Esys_CreatePrimary(esys_ctx, ESYS_TR_RH_OWNER,
+                            ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
```
single space at new line

idesai commented on tools/tpm2_load.c:
```diff
@@ -129,16 +337,87 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
+    
+    bool is_file_PEM = is_PEM(ctx.object.privpath);
+    if (is_file_PEM) {
+        BIO *input_bio = NULL;
+        input_bio = BIO_new_file(ctx.object.privpath, "rb");
```
```suggestion
        BIO *input_bio =  BIO_new_file(ctx.object.privpath, "rb");
```

idesai commented on tools/tpm2_load.c:
```diff
@@ -172,20 +451,24 @@ static tool_rc check_options(ESYS_CONTEXT *ectx) {
         LOG_ERR("Expected private object portion via -r");
         rc = tool_rc_option_error;
     }
+    // check for PEM file
+    // if not PEM then we need to make sure we check everything
```
```suggestion
    /* 
     * check for PEM file
     * if not PEM then we need to make sure we check everything
     */
```

idesai commented on tools/tpm2_load.c:
```diff
@@ -129,16 +337,87 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
+        } else {
+            // persistent parent
+            ESYS_TR tr_parent = ESYS_TR_NONE;
+            tool_rc tr_handle_rc = tpm2_from_tpm_public(ectx, parent_val,
```
```suggestion
            tool_rc rc = tpm2_from_tpm_public(ectx, parent_val,
```

idesai commented:
squash the commits

idesai committed:
tools/tpm2_ecdhzgen: Redo tools layout

Signed-off-by: Imran Desai <imran.desai@intel.com>

idesai committed:
tools/tpm2_ecephemeral: redo tools layout

Signed-off-by: Imran Desai <imran.desai@intel.com>

idesai committed:
tools/tpm2_geteccparameters: redo tools layout

Signed-off-by: Imran Desai <imran.desai@intel.com>

mgerstner committed:
tpm2_eventlog: simplify and optimize allocation for read loop

Currently the code allocates an unnecessary extra chunk. Simplify the
code and avoid memory waste.

Signed-off-by: Matthias Gerstner <matthias.gerstner@suse.de>

mayuanchenma committed:
add sm2 sign and verify test case.

Signed-off-by: mayuanchen <94815698+mayuanchenma@users.noreply.github.com>

ccihula committed:
tools/tpm2_ecdhkeygen: add support for calculating cpHash

Signed-off-by: cihula <connor.cihula@intel.com>

idesai committed:
authorizations.md: fix a typo

Fixes #3006 

Passwords specified as hex: must not preface a 0x

idesai committed:
tools/tpm2_fluscontext.c: Redo tools layout and fix segfault

Fixes #2402
Fixes #3015

Signed-off-by: Imran Desai <imran.desai@intel.com>
